# Test initialised fitting and sampling.

1. Explore whether fitting the SkillsGP model and using it as a starting point to fit a new match is faster.
2. Explore whether we can use a MAP-fitted skills vector as the initial sample for SkillsGPMCMC.

## Setup.

In [1]:
import os
import sys

os.chdir("..")

In [5]:
import importlib
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy.spatial.distance import squareform, pdist

from src import load, munge
import src.models.gp
import src.stats

In [530]:
matches = load.all_matches_df()
matches_ti9 = matches.loc[matches.league_name == "The International 2019"]
players_mat_ti9 = munge.make_match_players_matrix(
    matches_ti9.radiant_players, matches_ti9.dire_players)
players = src.munge.match_df_to_player_df(matches_ti9)

In [531]:
matches_ti9

startDate             league_name          radiant_name  \
matchId                                                                        
4888646940 2019-07-07 03:01:41  The International 2019               BOOM ID   
4888686136 2019-07-07 03:31:36  The International 2019                Adroit   
4888732113 2019-07-07 04:05:31  The International 2019             Team EVOS   
4888772755 2019-07-07 04:33:55  The International 2019            RESURGENCE   
4888859291 2019-07-07 05:20:41  The International 2019             Team EVOS   
4888936676 2019-07-07 06:01:24  The International 2019               Mineski   
4888974952 2019-07-07 06:22:32  The International 2019               BOOM ID   
4889038294 2019-07-07 06:53:26  The International 2019        Team Jinesbrus   
4889115579 2019-07-07 07:36:33  The International 2019               BOOM ID   
4889163811 2019-07-07 07:59:04  The International 2019               Mineski   
4889233900 2019-07-07 08:40:15  The International 2019            RESURGENCE   
4889265404 2019-07-07 08:59:58  The International 2019               BOOM ID   
4889343439 2019-07-07 09:46:20  The International 2019                AMPLFY   
4889360849 2019-07-07 09:57:22  The International 2019            RESURGENCE   
4889341660 2019-07-07 10:00:41  The International 2019         Natus Vincere   
4889414157 2019-07-07 10:34:44  The International 2019           Team Empire   
4889464422 2019-07-07 11:06:20  The International 2019        Winstrike Team   
4889496502 2019-07-07 11:24:49  The International 2019               Mineski   
4889502467 2019-07-07 11:28:55  The International 2019               BOOM ID   
4889525518 2019-07-07 11:42:02  The International 2019         Nemiga Gaming   
4889555136 2019-07-07 12:05:32  The International 2019        Winstrike Team   
4889610539 2019-07-07 12:27:52  The International 2019            496 Gaming   
4889635707 2019-07-07 12:40:50  The International 2019        Gambit Esports   
4889691529 2019-07-07 13:12:21  The International 2019         Natus Vincere   
4889722230 2019-07-07 13:29:16  The International 2019           Team Spirit   
4889865830 2019-07-07 14:32:27  The International 2019         Natus Vincere   
4889881001 2019-07-07 14:39:40  The International 2019           Team Empire   
4889981518 2019-07-07 15:23:11  The International 2019         Vega Squadron   
4889995626 2019-07-07 15:30:24  The International 2019           Team Spirit   
4890122311 2019-07-07 16:43:43  The International 2019         Nemiga Gaming   
...                        ...                     ...                   ...   
4901646019 2019-07-13 16:54:38  The International 2019  Unknown Radiant Team   
4901728834 2019-07-13 17:53:32  The International 2019    Chaos Esports Club   
4901733286 2019-07-13 17:57:51  The International 2019            Kookaburra   
4901742097 2019-07-13 18:04:37  The International 2019            beastcoast   
4901799568 2019-07-13 18:49:22  The International 2019        Forward Gaming   
4901835876 2019-07-13 19:13:43  The International 2019            beastcoast   
4901925673 2019-07-13 20:37:17  The International 2019            beastcoast   
4901998619 2019-07-13 21:37:01  The International 2019            beastcoast   
4902056634 2019-07-13 22:38:34  The International 2019            Kookaburra   
4902122478 2019-07-14 00:19:57  The International 2019               J.Storm   
4902186317 2019-07-14 01:39:53  The International 2019            Kookaburra   
4902354217 2019-07-14 03:59:21  The International 2019   Royal Never Give Up   
4902464697 2019-07-14 05:15:39  The International 2019                 EHOME   
4902576905 2019-07-14 06:29:03  The International 2019   Royal Never Give Up   
4902729928 2019-07-14 07:50:22  The International 2019                 CDEC    
4902837041 2019-07-14 08:44:30  The International 2019                 CDEC    
4902924459 2019-07-14 09:37:49  The International 2019                 CDEC    
4903051772 2019-07-14

## First fit a MAP model.

In [532]:
importlib.reload(src.models.gp)

two_years_in_ms = 2 * 365 * 24 * 60 * 60 * 1000
gp_map = src.models.gp.SkillsGPMAP(
    None,
    None,
    players_mat_ti9,
    matches_ti9.startTimestamp.values,
    matches_ti9.radiantVictory,
    "exponential",
    {"scale": two_years_in_ms},
    radi_prior_sd=100,  # Non-informative prior
    logistic_scale=1
)

In [533]:
gp_map.fit()

     fun: -7251.0131626120465
     jac: array([ 1.03590425e-05, -1.17897471e-05, -1.97483157e-06, ...,
        2.08626483e-05, -9.67426324e-06,  1.05673780e-05])
 message: 'Optimization terminated successfully.'
    nfev: 18
    nhev: 2119
     nit: 16
    njev: 33
  status: 0
 success: True
       x: array([-0.28478629, -0.28515649, -0.28527434, ..., -0.16480331,
       -0.16469054, -0.08426577])


In [534]:
gp_map.skills_mat(gp_map.fitted.x[:-1]).iloc[-1].dropna()

6922000      0.596078
10366616     0.596078
84429681     0.305941
86725175     0.305941
86726887     0.596078
87177591     0.305941
87196890     0.305941
94786276     0.305941
108452107    0.596078
221666230    0.596078
Name: 4904311356, dtype: float64

It took us 16 iterations and 2144 Hessian evaluations.

## Fit all but the last match, then add the last match.

In [535]:
matches_to_minus1 = matches_ti9.iloc[:-1]
players_mat_to_minus1 = munge.make_match_players_matrix(
    matches_to_minus1.radiant_players, matches_to_minus1.dire_players)
players_to_minus1 = src.munge.match_df_to_player_df(matches_to_minus1)

matches_minus1 = matches_ti9.iloc[[-1]]
players_mat_minus1 = munge.make_match_players_matrix(
    matches_minus1.radiant_players, matches_minus1.dire_players)
players_minus1 = src.munge.match_df_to_player_df(matches_minus1)

In [536]:
importlib.reload(src.models.gp)

two_years_in_ms = 2 * 365 * 24 * 60 * 60 * 1000
gp_map2 = src.models.gp.SkillsGPMAP(
    None,
    None,
    players_mat_to_minus1,
    matches_to_minus1.startTimestamp.values,
    matches_to_minus1.radiantVictory,
    "exponential",
    {"scale": two_years_in_ms},
    radi_prior_sd=100,
    logistic_scale=1
)

In [537]:
gp_map2.fit()

     fun: -7215.896591841696
     jac: array([-4.42373891e-06,  2.27665835e-06, -6.43913576e-06, ...,
        2.74491658e-06, -6.37987750e-06,  2.63498046e-05])
 message: 'Optimization terminated successfully.'
    nfev: 21
    nhev: 2561
     nit: 18
    njev: 38
  status: 0
 success: True
       x: array([-0.2846862 , -0.28505657, -0.28517459, ..., -0.16446207,
       -0.16434913, -0.07946344])


Idea:

1. Update the players matrix.
2. Check if the players of the new match are already fitted. If so, take the latest value. Otherwise use 0.0.
3. Convert the skills vector into a skills matrix.

### Try `predict()` and `predict_matches()`.

In [538]:
matches_minus1

startDate             league_name radiant_name  \
matchId                                                               
4904311356 2019-07-15 01:30:53  The International 2019      J.Storm   

                 dire_name  radiantVictory  \
matchId                                      
4904311356  Forward Gaming           False   

                              radiant_nicknames  \
matchId                                           
4904311356  [Moo, Resolut1on, Fear, 1437, nine]   

                                     dire_nicknames  \
matchId                                               
4904311356  [pieliedie, Sneyking, MSS, YawaR, CCnC]   

                                                 dire_players  dire_valveId  \
matchId                                                                       
4904311356  [6922000, 10366616, 86726887, 108452107, 22166...       6214538   

            duration  league_id  \
matchId                           
4904311356      1964      10749   

                                              radiant_players  \
matchId                                                         
4904311356  [84429681, 86725175, 87177591, 87196890, 94786...   

            radiant_valveId  seriesId  startTimestamp  
matchId                                                
4904311356          6288801    348164   1563154253000

In [539]:
matches_minus1.dire_players.iloc[0] + matches_minus1.radiant_players.iloc[0]

[6922000,
 10366616,
 86726887,
 108452107,
 221666230,
 84429681,
 86725175,
 87177591,
 87196890,
 94786276]

In [540]:
list(zip((matches_to_minus1.radiant_players + matches_to_minus1.dire_players),
          matches_to_minus1.startTimestamp))[:2]

[([152859296,
   156328257,
   237325411,
   301750126,
   389033587,
   87012746,
   91443418,
   113457795,
   192914280,
   412753955],
  1562468501000),
 ([100616105,
   104512126,
   113874152,
   127077041,
   355168766,
   3940262,
   88933594,
   112377459,
   126238768,
   375507918],
  1562470296000)]

In [541]:
pred_mat, var_mat = gp_map2.predict(
    [1562468501000],
    (matches_minus1.radiant_players.iloc[0]
     + matches_minus1.dire_players.iloc[0]))
display(pred_mat)
var_mat

84429681   86725175   87177591   87196890   94786276   \
1562468501000   0.312868   0.312868   0.312868   0.312868   0.312868   

               6922000    10366616   86726887   108452107  221666230  
1562468501000   0.575613   0.575613   0.575613   0.575613   0.575613

84429681   86725175   87177591   87196890   94786276   \
1562468501000   0.012488   0.012488   0.012488   0.012488   0.012488   

               6922000    10366616   86726887   108452107  221666230  
1562468501000   0.012488   0.012488   0.012488   0.012488   0.012488

In [542]:
display(matches_minus1)
gp_map2.predict_matches(matches_minus1.radiant_players,
                        matches_minus1.dire_players,
                        matches_minus1.startTimestamp.values)

startDate             league_name radiant_name  \
matchId                                                               
4904311356 2019-07-15 01:30:53  The International 2019      J.Storm   

                 dire_name  radiantVictory  \
matchId                                      
4904311356  Forward Gaming           False   

                              radiant_nicknames  \
matchId                                           
4904311356  [Moo, Resolut1on, Fear, 1437, nine]   

                                     dire_nicknames  \
matchId                                               
4904311356  [pieliedie, Sneyking, MSS, YawaR, CCnC]   

                                                 dire_players  dire_valveId  \
matchId                                                                       
4904311356  [6922000, 10366616, 86726887, 108452107, 22166...       6214538   

            duration  league_id  \
matchId                           
4904311356      1964      10749   

                                              radiant_players  \
matchId                                                         
4904311356  [84429681, 86725175, 87177591, 87196890, 94786...   

            radiant_valveId  seriesId  startTimestamp  
matchId                                                
4904311356          6288801    348164   1563154253000

pred_win_prob  win_prob-2sd  win_prob+2sd  radi_skill  \
1563154253000       0.200174      0.188626      0.212244    1.578399   

               radi_skill_sd  dire_skill  dire_skill_sd  radi_adv  
1563154253000       0.026076    2.884142       0.026076 -0.079463

### Manually create a new SkillsGPMAP object.

In [543]:
fitted_skills_mat = gp_map2.skills_mat(gp_map2.fitted.x[:-1])
fitted_skills_mat

3916428    3940262    4281729    6922000    10366616   11550182   \
matchId                                                                        
4888646940        NaN        NaN        NaN        NaN        NaN        NaN   
4888686136        NaN   0.197878        NaN        NaN        NaN        NaN   
4888732113        NaN        NaN        NaN        NaN        NaN        NaN   
4888772755        NaN        NaN        NaN        NaN        NaN        NaN   
4888859291        NaN        NaN        NaN        NaN        NaN        NaN   
4888936676        NaN        NaN        NaN        NaN        NaN        NaN   
4888974952        NaN        NaN        NaN        NaN        NaN        NaN   
4889038294        NaN   0.197790        NaN        NaN        NaN        NaN   
4889115579        NaN        NaN        NaN        NaN        NaN        NaN   
4889163811        NaN        NaN        NaN        NaN        NaN        NaN   
4889233900        NaN        NaN        NaN        NaN        NaN        NaN   
4889265404        NaN   0.197679        NaN        NaN        NaN        NaN   
4889343439        NaN        NaN        NaN        NaN        NaN        NaN   
4889360849        NaN        NaN        NaN        NaN        NaN        NaN   
4889341660        NaN        NaN        NaN        NaN        NaN        NaN   
4889414157        NaN        NaN        NaN        NaN        NaN  -0.285131   
4889464422        NaN        NaN        NaN        NaN        NaN        NaN   
4889496502        NaN   0.197491        NaN        NaN        NaN        NaN   
4889502467        NaN        NaN        NaN        NaN        NaN        NaN   
4889525518        NaN        NaN        NaN        NaN        NaN        NaN   
4889555136        NaN        NaN        NaN        NaN        NaN        NaN   
4889610539        NaN        NaN        NaN        NaN        NaN        NaN   
4889635707        NaN        NaN        NaN        NaN        NaN        NaN   
4889691529        NaN        NaN        NaN        NaN        NaN        NaN   
4889722230        NaN        NaN        NaN        NaN        NaN  -0.285110   
4889865830        NaN        NaN        NaN        NaN        NaN        NaN   
4889881001        NaN        NaN        NaN        NaN        NaN        NaN   
4889981518        NaN        NaN        NaN        NaN        NaN        NaN   
4889995626        NaN        NaN        NaN        NaN        NaN  -0.285248   
4890122311        NaN        NaN        NaN        NaN        NaN        NaN   
...               ...        ...        ...        ...        ...        ...   
4901517396        NaN        NaN        NaN        NaN        NaN        NaN   
4901646019        NaN        NaN        NaN        NaN        NaN        NaN   
4901728834        NaN        NaN        NaN        NaN        NaN        NaN   
4901733286        NaN        NaN        NaN   0.578788   0.578788        NaN   
4901742097        NaN        NaN        NaN        NaN        NaN        NaN   
4901799568        NaN        NaN        NaN   0.578762   0.578762        NaN   
4901835876        NaN        NaN        NaN        NaN        NaN        NaN   
4901925673        NaN        NaN        NaN   0.578692   0.578692        NaN   
4901998619        NaN        NaN        NaN   0.578637   0.578637        NaN   
4902056634        NaN        NaN        NaN        NaN        NaN        NaN   
4902122478        NaN        NaN        NaN        NaN        NaN        NaN   
4902186317        NaN        NaN        NaN        NaN        NaN        NaN   
4902354217        NaN        NaN        NaN        NaN        NaN        NaN   
4902464697        NaN        NaN        NaN        NaN        NaN        NaN   
4902576905        NaN        NaN        NaN        NaN        NaN        NaN   
4902729928        NaN        NaN        NaN        NaN        NaN        NaN   
4902837041        NaN        NaN        NaN        NaN        NaN        NaN   
4902924459  

In [544]:
def last_value_or_0(skills_mat, col_idx, default=0.0):
    """Get the last valid value or the default value in series s."""
    if col_idx not in skills_mat.columns:
        return default
    else:
        s = skills_mat[col_idx]
        idx = s.last_valid_index()
        if idx is None:
            return default
        else:
            return s[idx]

In [545]:
fitted_skills_mat.tail()

3916428    3940262    4281729    6922000    10366616   11550182   \
matchId                                                                        
4903974414        NaN        NaN        NaN        NaN        NaN        NaN   
4904062522        NaN        NaN        NaN        NaN        NaN        NaN   
4904171017        NaN        NaN        NaN   0.576858   0.576858        NaN   
4904218149        NaN        NaN        NaN   0.576875   0.576875        NaN   
4904261164        NaN        NaN        NaN   0.576868   0.576868        NaN   

            12231202   18180970   20321748   26316691   ...  352545711  \
matchId                                                 ...              
4903974414        NaN        NaN        NaN        NaN  ...        NaN   
4904062522        NaN        NaN        NaN        NaN  ...        NaN   
4904171017        NaN        NaN        NaN        NaN  ...        NaN   
4904218149        NaN        NaN        NaN        NaN  ...        NaN   
4904261164        NaN        NaN        NaN        NaN  ...        NaN   

            355168766  375507918  389033587  407321629  412753955  412758827  \
matchId                                                                        
4903974414        NaN        NaN        NaN        NaN        NaN        NaN   
4904062522        NaN        NaN        NaN        NaN        NaN        NaN   
4904171017        NaN        NaN        NaN        NaN        NaN        NaN   
4904218149        NaN        NaN        NaN        NaN        NaN        NaN   
4904261164        NaN        NaN        NaN        NaN        NaN        NaN   

            468599935  480412663  893135754  
matchId                                      
4903974414        NaN        NaN        NaN  
4904062522        NaN        NaN        NaN  
4904171017        NaN        NaN        NaN  
4904218149        NaN        NaN        NaN  
4904261164        NaN        NaN        NaN  

[5 rows x 240 columns]

In [546]:
imputed_skills = pd.Series([last_value_or_0(fitted_skills_mat, i)
                            for i in players_mat_minus1.columns],
                           index=players_mat_minus1.columns,
                           name=players_mat_minus1.index[0])
pd.DataFrame(imputed_skills).T

6922000    10366616   84429681   86725175   86726887   87177591   \
4904311356   0.576868   0.576868   0.315701   0.315701   0.576868   0.315701   

            87196890   94786276   108452107  221666230  
4904311356   0.315701   0.315701   0.576868   0.576868

In [547]:
new_skills_mat = fitted_skills_mat.append(pd.DataFrame(imputed_skills).T)
new_skills_mat.tail()

3916428    3940262    4281729    6922000    10366616   11550182   \
4904062522        NaN        NaN        NaN        NaN        NaN        NaN   
4904171017        NaN        NaN        NaN   0.576858   0.576858        NaN   
4904218149        NaN        NaN        NaN   0.576875   0.576875        NaN   
4904261164        NaN        NaN        NaN   0.576868   0.576868        NaN   
4904311356        NaN        NaN        NaN   0.576868   0.576868        NaN   

            12231202   18180970   20321748   26316691   ...  352545711  \
4904062522        NaN        NaN        NaN        NaN  ...        NaN   
4904171017        NaN        NaN        NaN        NaN  ...        NaN   
4904218149        NaN        NaN        NaN        NaN  ...        NaN   
4904261164        NaN        NaN        NaN        NaN  ...        NaN   
4904311356        NaN        NaN        NaN        NaN  ...        NaN   

            355168766  375507918  389033587  407321629  412753955  412758827  \
4904062522        NaN        NaN        NaN        NaN        NaN        NaN   
4904171017        NaN        NaN        NaN        NaN        NaN        NaN   
4904218149        NaN        NaN        NaN        NaN        NaN        NaN   
4904261164        NaN        NaN        NaN        NaN        NaN        NaN   
4904311356        NaN        NaN        NaN        NaN        NaN        NaN   

            468599935  480412663  893135754  
4904062522        NaN        NaN        NaN  
4904171017        NaN        NaN        NaN  
4904218149        NaN        NaN        NaN  
4904261164        NaN        NaN        NaN  
4904311356        NaN        NaN        NaN  

[5 rows x 240 columns]

In [548]:
players_mat_minus1

6922000    10366616   84429681   86725175   86726887   87177591   \
matchId                                                                        
4904311356       -1.0       -1.0        1.0        1.0       -1.0        1.0   

            87196890   94786276   108452107  221666230  
matchId                                                 
4904311356        1.0        1.0       -1.0       -1.0

In [549]:
new_players_mat = gp_map2.players_mat.append(players_mat_minus1).fillna(0.0)
new_players_mat

3916428    3940262    4281729    6922000    10366616   11550182   \
matchId                                                                        
4888646940        0.0        0.0        0.0        0.0        0.0        0.0   
4888686136        0.0       -1.0        0.0        0.0        0.0        0.0   
4888732113        0.0        0.0        0.0        0.0        0.0        0.0   
4888772755        0.0        0.0        0.0        0.0        0.0        0.0   
4888859291        0.0        0.0        0.0        0.0        0.0        0.0   
4888936676        0.0        0.0        0.0        0.0        0.0        0.0   
4888974952        0.0        0.0        0.0        0.0        0.0        0.0   
4889038294        0.0        1.0        0.0        0.0        0.0        0.0   
4889115579        0.0        0.0        0.0        0.0        0.0        0.0   
4889163811        0.0        0.0        0.0        0.0        0.0        0.0   
4889233900        0.0        0.0        0.0        0.0        0.0        0.0   
4889265404        0.0       -1.0        0.0        0.0        0.0        0.0   
4889343439        0.0        0.0        0.0        0.0        0.0        0.0   
4889360849        0.0        0.0        0.0        0.0        0.0        0.0   
4889341660        0.0        0.0        0.0        0.0        0.0        0.0   
4889414157        0.0        0.0        0.0        0.0        0.0       -1.0   
4889464422        0.0        0.0        0.0        0.0        0.0        0.0   
4889496502        0.0       -1.0        0.0        0.0        0.0        0.0   
4889502467        0.0        0.0        0.0        0.0        0.0        0.0   
4889525518        0.0        0.0        0.0        0.0        0.0        0.0   
4889555136        0.0        0.0        0.0        0.0        0.0        0.0   
4889610539        0.0        0.0        0.0        0.0        0.0        0.0   
4889635707        0.0        0.0        0.0        0.0        0.0        0.0   
4889691529        0.0        0.0        0.0        0.0        0.0        0.0   
4889722230        0.0        0.0        0.0        0.0        0.0        1.0   
4889865830        0.0        0.0        0.0        0.0        0.0        0.0   
4889881001        0.0        0.0        0.0        0.0        0.0        0.0   
4889981518        0.0        0.0        0.0        0.0        0.0        0.0   
4889995626        0.0        0.0        0.0        0.0        0.0        1.0   
4890122311        0.0        0.0        0.0        0.0        0.0        0.0   
...               ...        ...        ...        ...        ...        ...   
4901646019        0.0        0.0        0.0        0.0        0.0        0.0   
4901728834        0.0        0.0        0.0        0.0        0.0        0.0   
4901733286        0.0        0.0        0.0       -1.0       -1.0        0.0   
4901742097        0.0        0.0        0.0        0.0        0.0        0.0   
4901799568        0.0        0.0        0.0        1.0        1.0        0.0   
4901835876        0.0        0.0        0.0        0.0        0.0        0.0   
4901925673        0.0        0.0        0.0       -1.0       -1.0        0.0   
4901998619        0.0        0.0        0.0       -1.0       -1.0        0.0   
4902056634        0.0        0.0        0.0        0.0        0.0        0.0   
4902122478        0.0        0.0        0.0        0.0        0.0        0.0   
4902186317        0.0        0.0        0.0        0.0        0.0        0.0   
4902354217        0.0        0.0        0.0        0.0        0.0        0.0   
4902464697        0.0        0.0        0.0        0.0        0.0        0.0   
4902576905        0.0        0.0        0.0        0.0        0.0        0.0   
4902729928        0.0        0.0        0.0        0.0        0.0        0.0   
4902837041        0.0        0.0        0.0        0.0        0.0        0.0   
4902924459        0.0        0.0        0.0        0.0        0.0        0.0   
4903051772  

### Actually create the new object with the computed values.

In [550]:
importlib.reload(src.models.gp)

new_initial_skills = new_skills_mat.values.T[new_players_mat.values.T != 0.0]

two_years_in_ms = 2 * 365 * 24 * 60 * 60 * 1000
gp_map2_prime = src.models.gp.SkillsGPMAP(
    new_initial_skills,
    gp_map2.fitted.x[-1],
    new_players_mat,
    matches_ti9.startTimestamp.values,
    matches_ti9.radiantVictory,
    "exponential",
    {"scale": two_years_in_ms},
    radi_prior_sd=100,
    logistic_scale=1
)

In [551]:
gp_map2_prime.fit()

     fun: -7251.013162612773
     jac: array([ 6.18084507e-06, -8.36016304e-06, -5.12338929e-06, ...,
       -8.23186910e-06, -2.32293850e-06, -5.05975558e-05])
 message: 'Optimization terminated successfully.'
    nfev: 9
    nhev: 1163
     nit: 8
    njev: 16
  status: 0
 success: True
       x: array([-0.28478576, -0.28515596, -0.28527381, ..., -0.16480323,
       -0.16469046, -0.08426584])


By imputing the skills of the new samples, we get the following speedup:

* Time: from 1160 ms to 410 ms (65% decrease).
* Iterations: from 16 to 5 (68% decrease).
* Hessian evaluations: from 2144 to 678 (68% decrease).
* Objective: from -7251.013162611372 to -7251.013158887112.

In [552]:
gp_map2_prime.skills_mat(gp_map2_prime.fitted.x[:-1]).iloc[-1].dropna()

6922000      0.596078
10366616     0.596078
84429681     0.305941
86725175     0.305941
86726887     0.596078
87177591     0.305941
87196890     0.305941
94786276     0.305941
108452107    0.596078
221666230    0.596078
Name: 4904311356, dtype: float64

For references, the results from the full fitting are here.

    87012746     0.208980
    91443418     0.208980
    113457795    0.208980
    152859296   -0.036104
    156328257   -0.036104
    192914280    0.208980
    237325411   -0.036104
    301750126   -0.036104
    389033587   -0.036104
    412753955    0.208980
    Name: 4888646940, dtype: float64

In [553]:
gp_map2_prime.win_prob(0.005)

0.5012499973958399

### Incremental fitting based on `SkillsGPMAP.add_matches()`.

In [562]:
matches_minus1

startDate             league_name radiant_name  \
matchId                                                               
4904311356 2019-07-15 01:30:53  The International 2019      J.Storm   

                 dire_name  radiantVictory  \
matchId                                      
4904311356  Forward Gaming           False   

                              radiant_nicknames  \
matchId                                           
4904311356  [Moo, Resolut1on, Fear, 1437, nine]   

                                     dire_nicknames  \
matchId                                               
4904311356  [pieliedie, Sneyking, MSS, YawaR, CCnC]   

                                                 dire_players  dire_valveId  \
matchId                                                                       
4904311356  [6922000, 10366616, 86726887, 108452107, 22166...       6214538   

            duration  league_id  \
matchId                           
4904311356      1964      10749   

                                              radiant_players  \
matchId                                                         
4904311356  [84429681, 86725175, 87177591, 87196890, 94786...   

            radiant_valveId  seriesId  startTimestamp  
matchId                                                
4904311356          6288801    348164   1563154253000

In [563]:
importlib.reload(src.models.gp)

two_years_in_ms = 2 * 365 * 24 * 60 * 60 * 1000
gp_map2 = src.models.gp.SkillsGPMAP(
    None,
    None,
    players_mat_to_minus1,
    matches_to_minus1.startTimestamp.values,
    matches_to_minus1.radiantVictory,
    "exponential",
    {"scale": two_years_in_ms},
    radi_prior_sd=100,
    logistic_scale=1
)

In [564]:
gp_map2.fit()

     fun: -7215.896591841696
     jac: array([-4.42373891e-06,  2.27665835e-06, -6.43913576e-06, ...,
        2.74491658e-06, -6.37987750e-06,  2.63498046e-05])
 message: 'Optimization terminated successfully.'
    nfev: 21
    nhev: 2561
     nit: 18
    njev: 38
  status: 0
 success: True
       x: array([-0.2846862 , -0.28505657, -0.28517459, ..., -0.16446207,
       -0.16434913, -0.07946344])


In [565]:
gp_map2_prime = gp_map2.add_matches(players_mat_minus1,
                                    matches_minus1.startTimestamp,
                                    matches_minus1.radiantVictory)

In [566]:
gp_map2_prime.fit()

     fun: -7251.013162612773
     jac: array([ 6.18084507e-06, -8.36016304e-06, -5.12338929e-06, ...,
       -8.23186910e-06, -2.32293850e-06, -5.05975558e-05])
 message: 'Optimization terminated successfully.'
    nfev: 9
    nhev: 1163
     nit: 8
    njev: 16
  status: 0
 success: True
       x: array([-0.28478576, -0.28515596, -0.28527381, ..., -0.16480323,
       -0.16469046, -0.08426584])


### Compare the fitting results from obtained from full vs incremental fit.

In [574]:
gp_map.skills_mat(gp_map.fitted.x[:-1]).iloc[-1].dropna()

6922000      0.596078
10366616     0.596078
84429681     0.305941
86725175     0.305941
86726887     0.596078
87177591     0.305941
87196890     0.305941
94786276     0.305941
108452107    0.596078
221666230    0.596078
Name: 4904311356, dtype: float64

In [575]:
gp_map2_prime.skills_mat(gp_map2_prime.fitted.x[:-1]).iloc[-1].dropna()

6922000      0.596078
10366616     0.596078
84429681     0.305941
86725175     0.305941
86726887     0.596078
87177591     0.305941
87196890     0.305941
94786276     0.305941
108452107    0.596078
221666230    0.596078
Name: 4904311356, dtype: float64

## Test incremental fitting, but now using the full dataset.

In [ ]:
players_mat = munge.make_match_players_matrix(
    matches.radiant_players, matches.dire_players)
players = src.munge.match_df_to_player_df(matches)

In [ ]:
matches_to_minus1 = matches.iloc[:-1]
players_mat_to_minus1 = munge.make_match_players_matrix(
    matches_to_minus1.radiant_players, matches_to_minus1.dire_players)
players_to_minus1 = src.munge.match_df_to_player_df(matches_to_minus1)

matches_minus1 = matches.iloc[[-1]]
players_mat_minus1 = munge.make_match_players_matrix(
    matches_minus1.radiant_players, matches_minus1.dire_players)
players_minus1 = src.munge.match_df_to_player_df(matches_minus1)

In [ ]:
importlib.reload(src.models.gp)

two_years_in_ms = 2 * 365 * 24 * 60 * 60 * 1000
gp_map3 = src.models.gp.SkillsGPMAP(
    None,
    None,
    players_mat_to_minus1,
    matches_to_minus1.startTimestamp.values,
    matches_to_minus1.radiantVictory,
    players.loc[players_mat_to_minus1.columns].name,
    "exponential",
    {"scale": two_years_in_ms},
    radi_prior_sd=100,
    logistic_scale=1
)

In [ ]:
gp_map3.fit()

### Now perform the incremental fit.

In [84]:
matches.iloc[-1]

dire_name                                                   Rock.Young
dire_nicknames                        [Jixing, Gintoki, Ayo, 垃圾君, Lin]
dire_players         [92590479, 101908305, 110228930, 116852311, 13...
dire_valveId                                                   4288603
duration                                                          1566
league_id                                                         9601
league_name                                               EPICENTER XL
radiantVictory                                                    True
radiant_name                                               VGJ Thunder
radiant_nicknames                     [Sylar, ddc, Freeze, Yang, Fade]
radiant_players      [108382060, 114239371, 137272985, 139937922, 1...
radiant_valveId                                                5027210
seriesId                                                        209169
startDate                                          2018-03-23 08:10:59
startT

In [89]:
fitted_skills_mat = gp_map3.skills_mat(gp_map3.fitted.x[:-1])
imputed_skills = pd.Series([last_value_or_0(fitted_skills_mat, i)
                            for i in players_mat_minus1.columns],
                           index=players_mat_minus1.columns,
                           name=players_mat_minus1.index[0])
pd.DataFrame(imputed_skills).T

92590479   101908305  108382060  110228930  114239371  116852311  \
3795410349  -0.158531   0.242847   0.187159    0.28641   0.192424   0.242847   

            136477860  137272985  139937922  182331313  
3795410349   0.369735  -0.041949   0.214477   0.742586

In [90]:
new_skills_mat = fitted_skills_mat.append(pd.DataFrame(imputed_skills).T)

new_players_mat = gp_map3.players_mat.append(players_mat_minus1).fillna(0.0)

In [92]:
importlib.reload(src.models.gp)

new_initial_skills = new_skills_mat.values.T[new_players_mat.values.T != 0.0]

two_years_in_ms = 2 * 365 * 24 * 60 * 60 * 1000
gp_map3_prime = src.models.gp.SkillsGPMAP(
    new_initial_skills,
    gp_map3.fitted.x[-1],
    new_players_mat,
    matches.startTimestamp.values,
    matches.radiantVictory,
    players.loc[players_mat.columns].name,
    "exponential",
    {"scale": two_years_in_ms},
    radi_prior_sd=100,
    logistic_scale=1
)

In [93]:
gp_map3_prime.fit()

     fun: -140676.34710956446
     jac: array([ 3.50886888e-06,  2.86890205e-06, -7.95669708e-06, ...,
       -1.57840128e-06,  3.07009357e-07,  8.79081164e-05])
 message: 'Optimization terminated successfully.'
    nfev: 5
    nhev: 5198
     nit: 4
    njev: 8
  status: 0
 success: True
       x: array([0.74687263, 0.74696523, 0.74989353, ..., 0.10355485, 0.10417335,
       0.17807179])
